### ISSUE TITLE
VAXstation 4000 VLC takes inordinately long to boot to console prompt

### ISSUE THREAD

**Issue by patbarron** [simh/simh#1113]
- #### Context

When starting the VAXstation 4000 VLC simulator, and typing the "BOOT" command, it takes just over 2 hours (I timed it) for the simulator to reach the ">>>" prompt from the ROM bootstrap.  This behavior is different from all of the other SIMH VAX simulators I've looked at that include ROM monitor code, which reach the console prompt in a reasonable amount of time (comparable to what I recall experiencing on real hardware).

This issue appears to have something to do with the SIMH ROM delay function.  If I issue the "SET ROM NODELAY" command in SIMH before attempting to boot, the simulator reaches the ">>>" prompt in a reasonable amount of time (comparable to what I see in the other VAX simulators, though I do not have to use the "SET ROM NODELAY" command with those to get to a ROM monitor prompt in a reasonable amount of time).

- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue

```
sim> show version
VAXstation 4000-VLC (KA48) simulator V4.0-0 Current
    Simulator Framework Capabilities:
        64b data
        64b addresses
        Threaded Ethernet Packet transports:TAP:NAT:UDP
        Idle/Throttling support is available
        Virtual Hard Disk (VHD) support
        RAW disk and CD/DVD ROM support
        Asynchronous I/O support (Lock free asynchronous event queue)
        Asynchronous Clock support
        FrontPanel API Version 12
    Host Platform:
        Compiler: GCC 11.2.1 20211203 (Red Hat 11.2.1-7)
        Simulator Compiled as C arch: x64 (Release Build) on Jan 10 2022 at 15:55:39
        Build Tool: simh-makefile
        Memory Access: Little Endian
        Memory Pointer Size: 64 bits
        Large File (>2GB) support
        SDL Video support: No Video Support
        PCRE RegEx (Version 8.45 2021-06-15) support for EXPECT commands
        OS clock resolution: 1ms
        Time taken by msleep(1): 1ms
        OS: Linux _(hostname redacted)_ 5.15.12-200.fc35.x86_64 #1 SMP Wed Dec 29 15:03:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
        tar tool: tar (GNU tar) 1.34
        curl tool: curl 7.79.1 (x86_64-redhat-linux-gnu) libcurl/7.79.1 OpenSSL/1.1.1l-fips zlib/1.2.11 brotli/1.0.9 libidn2/2.3.2 libpsl/0.21.1 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.45.1 OpenLDAP/2.4.59
        git commit id: 7756a38e
        git commit time: 2022-01-10T04:08:44-0800
```
- #### how you built the simulator or that you're using prebuilt binaries

```
$ make vaxstation4000vlc
lib paths are: /lib/ /lib64/ /usr/lib64/dyninst/ /usr/lib64/iscsi/ /usr/lib64/llvm11/lib/ /usr/lib64/llvm12/lib/ /usr/lib64/pipewire-0.3/jack/ /usr/lib64/tcl8.6/ /usr/lib/
include paths are:  /usr/lib/gcc/x86_64-redhat-linux/11/include /usr/local/include /usr/include
using libm: /lib64/libm.so
using librt: /lib64/librt.a
using libpthread: /lib64/libpthread.a /usr/include/pthread.h
using libpcre: /lib64/libpcre.so /usr/include/pcre.h
using semaphore: /usr/include/semaphore.h
using libdl: /lib64/libdl.a /usr/include/dlfcn.h
using libpng: /lib64/libpng.so /usr/include/png.h
using zlib: /lib64/libz.so /usr/include/zlib.h
using mman: /usr/include/sys/mman.h
*** Warning ***
*** Warning *** vaxstation4000vlc Simulator is being built WITHOUT
*** Warning *** libpcap networking support
*** Warning ***
*** Warning *** To build simulator(s) with libpcap networking support you
*** Warning *** should read 0readme_ethernet.txt and follow the instructions
*** Warning *** regarding the needed libpcap development components for your
*** Warning *** Linux platform
*** Warning ***
*** Info ***
*** Info *** Simulators on your Linux platform can also be built with
*** Info *** extended LAN Ethernet networking support by using VDE Ethernet.
*** Info ***
*** Info *** To build simulator(s) with extended networking support you
*** Info *** should read 0readme_ethernet.txt and follow the instructions
*** Info *** regarding the needed libvdeplug components for your Linux
*** Info *** platform
*** Info ***
***
*** vaxstation4000vlc Simulator being built with:
*** - compiler optimizations and no debugging support. GCC Version: 11.2.1.
*** - Local LAN packet transports: TAP NAT(SLiRP)
*** - Per simulator tests will be run.
***
*** git commit id is 7756a38e315cd189619158443b84480a8c63a732.
*** git commit time is 2022-01-10T04:08:44-0800.
***
gcc -std=gnu99 -U__STRICT_ANSI__  -O2 -finline-functions -fgcse-after-reload -fpredictive-commoning -fipa-cp-clone -fno-unsafe-loop-optimizations -fno-strict-overflow -DSIM_GIT_COMMIT_ID=7756a38e315cd189619158443b84480a8c63a732 -DSIM_GIT_COMMIT_TIME=2022-01-10T04:08:44-0800  -DSIM_COMPILER="GCC Version: 11.2.1" -DSIM_BUILD_TOOL=simh-makefile -I . -D_GNU_SOURCE -DUSE_READER_THREAD -DSIM_ASYNCH_IO  -DHAVE_PCRE_H -DHAVE_SEMAPHORE -DHAVE_SYS_IOCTL -DHAVE_LINUX_CDROM -DSIM_HAVE_DLOPEN=so -DHAVE_UTIME -DHAVE_LIBPNG -DHAVE_ZLIB -DHAVE_GLOB -DHAVE_SHM_OPEN  ./VAX/vax_cpu.c ./VAX/vax_cpu1.c ./VAX/vax_fpa.c ./VAX/vax_cis.c ./VAX/vax_octa.c ./VAX/vax_cmode.c ./VAX/vax_mmu.c ./VAX/vax_sys.c ./VAX/vax_syscm.c ./VAX/vax_watch.c ./VAX/vax_nar.c ./VAX/vax4xx_stddev.c ./VAX/vax440_sysdev.c ./VAX/vax440_syslist.c ./VAX/vax4xx_dz.c ./VAX/vax_xs.c ./VAX/vax_lk.c ./VAX/vax_vs.c ./VAX/vax4xx_rz94.c ./sim_scsi.c ./scp.c ./sim_console.c ./sim_fio.c ./sim_timer.c ./sim_sock.c ./sim_tmxr.c ./sim_ether.c ./sim_tape.c ./sim_disk.c ./sim_serial.c ./sim_video.c ./sim_imd.c ./sim_card.c -DVM_VAX -DVAX_440 -DUSE_INT64 -DUSE_ADDR64 -I ./VAX -DHAVE_TAP_NETWORK -DUSE_NETWORK -Islirp -Islirp_glue -Islirp_glue/qemu -DHAVE_SLIRP_NETWORK -DUSE_SIMH_SLIRP_DEBUG slirp/*.c slirp_glue/*.c -DVAX_48 -o BIN/vaxstation4000vlc -lm -lrt -lpthread -lpcre -ldl -lpng -lz
BIN/vaxstation4000vlc RegisterSanityCheck /home/pbarron/src/simh/VAX/tests/vax-diag_test.ini </dev/null
 Running internal register sanity checks on VAXstation 4000-VLC (KA48) simulator.
*** Good Registers in VAXstation 4000-VLC (KA48) simulator.

VAXstation 4000-VLC (KA48) simulator V4.0-0 Current        git commit id: 7756a38e
No diagnostics are available for the VAXstation 4000-VLC (KA48) Simulator

$
```

- #### the simulator configuration file (or commands) which were used when the problem occurred.

This occurs without even using a configuration file - in the default configuration for the simulator when you first start it up.

This instance of SIMH is running under Fedora 35, on a VMware virtual machine provisioned on a VMware ESXi hypervisor.

- #### the expected behavior and the actual behavior

The output looks reasonable, the problem is the amount of time it takes to produce the output.  The expected behavior is that after typing the SIMH "BOOT" command, the simulator reaches the ">>>" prompt from the ROM monitor within a reasonable amount of time (less than a minute - really, less than 30 seconds, with the default 16 GB memory configuration).  The actual behavior is that it takes over 2 hours to get there.

This is the output of the simulator when using "SET ROM NODELAY" - the output without it is basically the same, except it takes over 2 hours to produce...

```
$ vaxstation4000vlc

VAXstation 4000-VLC (KA48) simulator V4.0-0 Current        git commit id: 7756a38e
sim> set rom nodelay
sim> boot
Loading boot code from internal ka48a.bin



KA48-A V1.2-32B-V4.0
08-00-2B-7A-F7-13
16MB
|                                     |
#######################################
?? 010   2       LCG  0086
?? 001   3        DZ  0032
?? 001   4     CACHE  0512
?? 001   7        IT  0514
?? 001   8       SYS  0128
?? 001   9        NI  0024

>>>

>>> 6c
```

(The newline and the "6c" characters were not typed by me, those are phantom characters that spontaneously appear - I will open a separate issue for this....)

**Comment by markpizz** [comment]
2 hours would have something to do with the speed of the host system you are running on.

On my new Laptop (11th Gen Intel(R) Core(TM) i7-11390H @ 3.40GHz   2.92 GHz) I just saw this:
> VAXstation 4000-VLC (KA48) simulator V4.0-0 Current        git commit id: 1cffbd57+uncommitted-changes
> sim> do -v booting.ini
> ../booting.ini-1> SET ROM DELAY
> ../booting.ini-2> expect ">>> " echo Done at 16:14:49
> ../booting.ini-3> echo Starting at 16:14:49
> Starting at 16:14:49
> ../booting.ini-4> boot
> Loading boot code from internal ka48a.bin
>
> ◄
> ◄
> KA48-A V1.2-32B-V4.0
> 08-00-2B-57-F8-02
> 16MB
> 
> ?? 010   2       LCG  0086
> ?? 001   3        DZ  0032
> ?? 001   4     CACHE  0512
> ?? 001   7        IT  1280
> ?? 001   8       SYS  0128
> ?? 001   9        NI  0024
> ◄
> >>>
> Expect matched, PC: 2004AEE0 (INCW 8(R2))
> ../booting.ini-4> echo Done at 16:16:39
> Done at 16:16:39
> sim>

That was on Windows, but under WSL I see 1 minute 47 seconds.

Testing with SET ROM NODELAY and it gets to the >>> prompt in about 4 seconds.

**Comment by patbarron** [comment]
This is the processor information for the hypervisor host.  Does this elapsed time make sense to you given this information?  The VM itself is configured for 4 virtual CPUs and 8 GB of RAM.

![image](https://user-images.githubusercontent.com/1931014/149643595-b7585705-32de-487e-a7ac-d464a577442d.png)

**Comment by markpizz** [comment]
I can't explain your performance issue, but maybe it is due to the size of the processor cache and the potential for trashing the cache while the ROM based memory tests are being performed.

**Comment by patbarron** [comment]
So, this is interesting...  I pulled the most recent code from GitHub and rebuilt, and I don't see the problem anymore.  Then, I checked out commit 7756a38e again, and rebuilt ... and still don't see the problem.

In the interim, I've also done some updates on my local system, including a kernel update, and I can't prove that something in my environment previously wasn't triggering this (and can't easily revert to check).

I'll go ahead and close this issue, since I can no longer reproduce the problem (which I'd previously been able to reproduce consistently, at the time I first opened this issue).

**Comment by patbarron** [comment]
OK, I may have spoken too soon, because I've seen this behavior again - but I can't reliably reproduce it.  And, when I did see it again, I rebooted the VM - and after reboot, the behavior is gone (at least for now).  So maybe this is some type of cache thrashing issue, or something of that nature.

I'll leave this closed for now - will reopen if I can find a way to reliably reproduce the problem, and at least am somewhat confident that it's not something related to my environment.
